sync/vm/registry_test.go

```
@@ -173,20 +174,179 @@ func TestSyncerRegistry_RunSyncerTasks(t *testing.T) {
 		})
 	}
 }
+
+func TestSyncerRegistry_RunSyncerTasks_Concurrency(t *testing.T) {
+	// Barrier pattern:
+	// 1. All barrier syncers begin Sync concurrently (via errgroup).
+	// 2. Each syncer signals readiness using allStartedWG.Done().
+	// 3. Each syncer blocks until `releaseCh` is closed.
+	// 4. The test waits on allStartedWG.Wait() to ensure all have started.
+	// 5. The test then closes `releaseCh` to release them simultaneously.
+	//
+	// This demonstrates true concurrency: if syncers ran sequentially, the test
+	// would hang because no syncer could complete to allow the next to begin.
+
+	type testCase struct {
+		name              string
+		numBarrierSyncers int
+		numErrorSyncers   int
+		numCancelSyncers  int
+		errorMsg          string
+	}
+
+	// Helper to determine if test should succeed.
+	shouldSucceed := func(tc testCase) bool {
+		return tc.numErrorSyncers == 0
+	}
+
+	tests := []testCase{
+		{
+			name:              "single syncer succeeds",
+			numBarrierSyncers: 1,
+		},
+		{
+			name:              "multiple syncers start concurrently",
+			numBarrierSyncers: 5,
+		},
+		{
+			name:              "error syncer cancels barrier syncers",
+			numBarrierSyncers: 2,
+			numErrorSyncers:   1,
+			numCancelSyncers:  1,
+			errorMsg:          "test error",
+		},
+		{
+			name:             "multiple error syncers - first one wins",
+			numErrorSyncers:  3,
+			numCancelSyncers: 2,
+			errorMsg:         "boom",
+		},
+		{
+			name:              "no syncers registered",
+			numBarrierSyncers: 0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			registry := NewSyncerRegistry()
+
+			// Timeouts are derived per-subtest from its deadline to provide granular,
+			// phase-specific failure messages while respecting `go test -timeout`.
+			ctx, cancel := utilstest.NewTestContext(t)
+			t.Cleanup(cancel)
+
+			// Derive phase budgets from the subtest's deadline (or fallback).
+			baseTimeout := 5 * time.Second
```

This is confusing. Why do we even need these timeouts?

---

Ensure tests don’t hang -> The concurrency tests use barriers, errgroup cancellation, and channel signaling. If a bug prevents a syncer from starting, a release signal from being sent, or cancellation from propagating, the test could deadlock indefinitely. The timeouts guarantee the test fails fast instead of hanging.

Why we have more than one timeout? Because of “all syncers started” and “tasks completed” phases. That yields clearer failure messages about which phase stalled, not just a global “test timed out”.